recipes/lora_dpo_single_device.py

```diff
@@ -0,0 +1,629 @@
+
+        # Construct the full state dict with LoRA weights merged into base LLM weights
+        merged_state_dict = get_merged_lora_ckpt(
+            state_dict,
+            rank=self._lora_rank,
+            alpha=self._lora_alpha,
+        )
+        ckpt_dict.update({utils.MODEL_KEY: merged_state_dict})
+
+        # Construct the adapter weights
+        adapter_key_filter = lambda x: x in self.adapter_params
+        adapter_state_dict = {
+            k: v for k, v in self._model.state_dict().items() if adapter_key_filter(k)
+        }
+        ckpt_dict.update({utils.ADAPTER_KEY: adapter_state_dict})
+        self._checkpointer.save_checkpoint(
+            ckpt_dict,
+            epoch=epoch,
+            intermediate_checkpoint=(epoch + 1 < self.total_epochs),
+        )
+
+    def _should_update_weights(self, current_iteration: int) -> bool:
+        """
+        Determines whether the weights should be updated on the current iteration or not.
+        True is returned either if we've accumulated gradients for enough steps.
+        """
+        should_update_weights = (
+            current_iteration + 1
+        ) % self._gradient_accumulation_steps == 0
+        return should_update_weights
+
+    def concatenated_forward(self, model, batch):
+        concatenated_batch = self.concatenated_inputs(
+            batch,
+            label_pad_token_id=self._loss_fn.ignore_index,
+            padding_value=self._tokenizer.pad_id,
+            device=self._device,
+        )
+        len_chosen = batch["chosen_labels"].shape[0]
+
+        all_logits = model(concatenated_batch['concatenated_input_ids'])
+
+        all_logps = self.get_batch_logps(
+            all_logits,concatenated_batch["concatenated_labels"],self._loss_fn.ignore_index)
+        
+
+        chosen_logps = all_logps[:len_chosen]
+        rejected_logps = all_logps[len_chosen:]
+
+        chosen_logits = all_logits[:len_chosen]
+        rejected_logits = all_logits[len_chosen:]
+
+        return (chosen_logps, rejected_logps, chosen_logits, rejected_logits)
+    
+    @staticmethod
+    def concatenated_inputs(
+        batch: Dict[str, Union[List, torch.LongTensor]],
+        label_pad_token_id: int = -100,
+        padding_value: int = 0,
+        device: Optional[torch.device] = None,
+    ) -> Dict[str, torch.LongTensor]:
+        concatenated_batch = {}
+        max_length = max(batch["chosen_input_ids"].shape[1], batch["rejected_input_ids"].shape[1])
+
+        for k in batch:
+            if k.startswith("chosen") and isinstance(batch[k], torch.Tensor):
+                if "labels" in k:
+                    pad_value = label_pad_token_id
+                elif k.endswith("_input_ids"):
+                    pad_value = padding_value
+
+                concatenated_key = k.replace("chosen", "concatenated")
+                concatenated_batch[concatenated_key] = utils.pad_to_length(batch[k], max_length, pad_value=pad_value)
```

Comment:
why are we padding again here? if the collater already has access to both `chosen_input_ids` and `rejected_input_ids`, can't this be handled in `padded_collate_dpo`?

Reply:
Sure! Actually, I have the same question before, I'll move this to 'padded_collate_dpo'.